firefox_puppeteer/ui/windows.py

```diff
@@ -0,0 +1,479 @@
+        """
+        return [self.create_window_instance(handle) for handle in
+                self.marionette.chrome_window_handles]
+
+    @property
+    def current(self):
+        """Retrieves the currently selected chrome window.
+
+        :returns: The :class:`BaseWindow` for the currently active window.
+        """
+        return self.create_window_instance(self.marionette.current_chrome_window_handle)
+
+    @property
+    def focused_chrome_window_handle(self):
+        """Returns the currently focused chrome window handle
+
+        :returns: The `window handle` of the focused chrome window.
+        """
+        with self.marionette.using_context('chrome'):
+            return self.marionette.execute_script("""
+              Cu.import("resource://gre/modules/Services.jsm");
+              var win = Services.wm.getMostRecentWindow("");
+              return win.QueryInterface(Ci.nsIInterfaceRequestor)
+                        .getInterface(Ci.nsIDOMWindowUtils)
+                        .outerWindowID.toString();
+            """)
+
+    def close(self, handle):
+        """Closes the chrome window with the given handle.
+
+        :param handle: The handle of the chrome window
+
+        :returns: TODO
+        """
+        self.switch_to(handle)
+
+        # TODO: Maybe needs to wait as handled via an observer
+        return self.marionette.close_chrome_window()
+
+    def close_all(self, exceptions=None):
+        """Closes all open chrome windows.
+
+        There is an optional `exceptions` list, which can be used to exclude
+        specific chrome windows from being closed.
+
+        :param exceptions: Optional, list of handles or :class:`BaseWindow` instances
+        not to close
+        """
+        handles_to_keep = exceptions or []
+        handles_to_keep = [entry.handle if isinstance(entry, BaseWindow) else entry
+                           for entry in exceptions]
+
+        for handle in self.marionette.chrome_window_handles:
+            if handle not in handles_to_keep:
+                self.close(handle)
+
+    def create_window_instance(self, handle, expected_class=None):
+        """Creates a :class:`BaseWindow` instance for the given chrome window
+
+        :param handle: The handle of the chrome window
+        :param expected_class: Optional, check for the correct window class
+        """
+        current_handle = self.marionette.current_chrome_window_handle
+
+        try:
+            # Retrieve window type to determine the type of window
+            self.switch_to(handle)
+            window_type = self.marionette.get_window_type()
+        finally:
+            # Ensure to switch back to the original window
+            if handle != current_handle:
+                self.marionette.switch_to_window(current_handle)
+
+        window = BaseWindow(lambda: self.marionette, handle)
+
+        if window_type == 'navigator:browser':
+            window.__class__ = BrowserWindow
```

Comment:
The test should basically fail for an unknown window. But I think we can do that. There is really no need to have an instance of a BaseWindow which is more like an abstract class. I will move the instantiation into the if block. Thanks.